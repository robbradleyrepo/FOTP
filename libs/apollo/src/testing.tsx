import { MockedResponse, MockLink } from "@apollo/react-testing";
import { ApolloClientType, NextPageWithApollo } from "@sss/apollo";
import { Locale } from "@sss/i18n";
import { getMockedRouter } from "@sss/next/testing";
import {
  act,
  render,
  RenderOptions,
  RenderResult,
} from "@testing-library/react";
import {
  InMemoryCache,
  InMemoryCacheConfig,
  IntrospectionFragmentMatcher,
  IntrospectionResultData,
} from "apollo-cache-inmemory";
import { ApolloClient } from "apollo-client";
import {
  GetServerSideProps,
  GetServerSidePropsContext,
  GetStaticProps,
} from "next";
import { RouterContext } from "next/dist/shared/lib/router-context";
import { Router } from "next/router";
import { ParsedUrlQuery, stringify } from "querystring";
import React, { ComponentType } from "react";

import { initializeApollo } from "./client";
jest.mock("./client");

interface BaseSandbox {
  // The global mocks to apply
  globalMocks?: MockedResponse[];
  // Additional mocks under test
  mocks?: MockedResponse[];
  renderOptions?: Omit<RenderOptions, "queries">;
  pathname?: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  params?: { [key: string]: any };
}

interface PageSandbox extends BaseSandbox {
  NextPage: {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    default: React.ComponentType<any>;
    getStaticProps?: GetStaticProps;
    getServerSideProps?: GetServerSideProps;
  };
}

interface PageResult {
  mockLink: MockLink;
  redirect?: { permanent: boolean; destination: string };
  notFound?: boolean;
  page: RenderResult;
  router: Router;
}

interface ComponentSandbox extends BaseSandbox {
  Component: React.ReactNode;
}

interface ComponentResult {
  mockLink: MockLink;
  component: RenderResult;
  router: Router;
}

interface SandboxInit {
  globalMocks?: MockedResponse[];
  introspectionResultData?: IntrospectionResultData;
  locale: Locale;
  Wrapper: ComponentType<{
    apolloClient: ApolloClientType;
    locale: Locale;
  }>;
}

export const createApolloClient = (
  mocks: MockedResponse[],
  introspectionQueryResultData?: IntrospectionResultData
) => {
  const cacheConfig: InMemoryCacheConfig = { addTypename: true };

  // `InMemoryCache` doesn't like `fragmentMatcher: undefined`. We must omit it.
  if (introspectionQueryResultData) {
    cacheConfig.fragmentMatcher = new IntrospectionFragmentMatcher({
      introspectionQueryResultData,
    });
  }

  return new ApolloClient({
    cache: new InMemoryCache(cacheConfig),
    link: new MockLink(mocks, true),
  });
};

export default class Sandbox {
  public constructor(private opts: SandboxInit) {}

  protected createApolloClient(
    mocks: MockedResponse[] = [],
    globalMocks: MockedResponse[] = this.opts.globalMocks ?? []
  ) {
    return createApolloClient(
      [...globalMocks, ...mocks],
      this.opts.introspectionResultData
    );
  }

  public async page({
    NextPage: NextPageBundle,
    globalMocks,
    mocks,
    renderOptions,
    pathname = "/",
    params = {},
  }: PageSandbox): Promise<PageResult> {
    const router = getMockedRouter(pathname, params);

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (global as any).window = Object.create(window);
    Object.defineProperty(window, "location", {
      value: {
        href: `https://localhost${router.asPath}`,
        pathname: router.asPath,
        search: params ? stringify(params) : "",
      },
      writable: true,
    });

    const apolloClient = this.createApolloClient(mocks, globalMocks);

    (initializeApollo as jest.Mock).mockReturnValueOnce(apolloClient);

    const { default: PageComponent } = NextPageBundle;

    const pageData = NextPageBundle.getStaticProps
      ? await NextPageBundle.getStaticProps({ locale: "en-US", params })
      : NextPageBundle.getServerSideProps
      ? await NextPageBundle.getServerSideProps(
          // `getServerSideProps` context argument also requires
          // `req: IncomingMessage` and `res: ServerResponse`. Cast
          // this for testing purposes.
          ({
            locale: "en-US",
            params,
            req: {},
            res: {
              setHeader: jest.fn(),
              statusCode: null,
            },
          } as unknown) as GetServerSidePropsContext<ParsedUrlQuery>
        )
      : { props: {} };

    let page;

    act(() => {
      const Wrapper = this.opts.Wrapper;
      page = render(
        "props" in pageData &&
          !("notFound" in pageData) &&
          !("redirect" in pageData) ? (
          <RouterContext.Provider value={router}>
            <Wrapper apolloClient={apolloClient} locale={this.opts.locale}>
              <PageComponent {...pageData.props} />
            </Wrapper>
          </RouterContext.Provider>
        ) : (
          <div />
        ),
        {
          ...renderOptions,
        }
      );
    });

    return {
      link: apolloClient.link,
      notFound: "notFound" in pageData && pageData.notFound,
      page,
      redirect: "redirect" in pageData && pageData.redirect,
      router,
    } as any; // eslint-disable-line @typescript-eslint/no-explicit-any
  }

  public async component({
    Component,
    ...options
  }: ComponentSandbox): Promise<ComponentResult> {
    const NextPageComponent: NextPageWithApollo = () => {
      return <>{Component}</>;
    };

    const { page: component, ...rest } = await this.page({
      NextPage: { default: NextPageComponent },
      ...options,
    });

    return {
      component,
      ...rest,
    };
  }
}
