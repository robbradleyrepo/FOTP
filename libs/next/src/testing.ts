import { Router } from "next/router";
import { ParsedUrlQuery } from "querystring";
import { mocked } from "ts-jest/utils";

import { getAsPath } from "./helpers";

export const getMockedRouter = (pathname: string, params?: ParsedUrlQuery) => {
  const asPath = getAsPath(pathname, params);

  return mocked(({
    asPath,
    events: {
      off: jest.fn(),
      on: jest.fn(),
    },
    isReady: true,
    pathname,
    prefetch: jest.fn(async () => undefined),
    push: jest.fn(),
    query: params,
    replace: jest.fn(),
    route: pathname,
  } as unknown) as Router);
};
