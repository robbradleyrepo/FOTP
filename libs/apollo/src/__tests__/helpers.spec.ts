import { createApolloClient } from "@sss/apollo/testing";
import { defaultDataIdFromObject } from "apollo-cache-inmemory";
import { ApolloError } from "apollo-client";
import { GraphQLError } from "graphql";
import gql from "graphql-tag";

import {
  dataIdFromObject,
  runServerSideQuery,
  sanitize,
  stringifyApolloError,
} from "../helpers";

const __typename = "Test";
const _id = "123";
const id = "abc";
const _meta = {
  id: "xyz",
};

describe("dataIdFromObject", () => {
  it("should return the value returned by `defaultDataIdFromObject` if the standard fields are available", () => {
    [
      { __typename, _id, _meta, id },
      { __typename, _id, _meta },
      { __typename, _meta, id },
    ].forEach((obj) => {
      // First we'll check that we've got a valid ID. While is dependent on
      // `defaultDataIdFromObject`, we'll likely want to update our
      // implementation if this changes.
      expect(dataIdFromObject(obj)).not.toBeNull();

      expect(dataIdFromObject(obj)).toBe(defaultDataIdFromObject(obj));
    });
  });

  it("should use the Prismic meta ID if other ID fields are not available", () => {
    expect(dataIdFromObject({ __typename, _meta })).toBe(
      `${__typename}:${_meta.id}`
    );
  });

  it("should return `null` if there is not sufficient data to generate a unique ID", () => {
    expect(dataIdFromObject({ __typename })).toBeNull();
    expect(dataIdFromObject({ _meta })).toBeNull();
    expect(dataIdFromObject({})).toBeNull();
  });

  it("should add a `:completed` suffix to completed checkouts", () => {
    const incompleteCheckout = {
      __typename: "RCheckout",
    };
    const completeCheckout = {
      ...incompleteCheckout,
      completedAt: new Date().toISOString(),
    };

    expect(dataIdFromObject(incompleteCheckout)).toBe(
      defaultDataIdFromObject(incompleteCheckout)
    );
    expect(dataIdFromObject(completeCheckout)).toBe(
      `${defaultDataIdFromObject(incompleteCheckout)}:completed`
    );
  });
});

describe("sanitize", () => {
  it("should remove all `__typename` keys", () => {
    expect(
      sanitize({
        bar: {
          __typename: "Bar",
          nested: {
            __typename: "Nested",
            value: 123,
          },
        },
        baz: "baz",
        foo: [
          { __typename: "Foo", index: 0 },
          { __typename: "Foo", index: 1 },
          { __typename: "Foo", index: 3 },
        ],
      })
    ).toMatchSnapshot();
  });
});

describe("stringifyApolloError", () => {
  const anonymizeError = <T extends Error>(error: T) => {
    error.stack = `Dummy \`${error.constructor.name ?? "Error"}\` stack`;
    return error;
  };

  it("should stringify all top-level error data", () => {
    const apolloError = anonymizeError(new ApolloError({}));

    expect(stringifyApolloError(apolloError)).toMatchSnapshot();
  });

  it("should stringify GraphQL errors", () => {
    const apolloError = anonymizeError(
      new ApolloError({
        graphQLErrors: [anonymizeError(new GraphQLError("Foo"))],
      })
    );

    expect(stringifyApolloError(apolloError)).toMatchSnapshot();
  });

  it("should stringify network errors", () => {
    const apolloError = anonymizeError(
      new ApolloError({
        networkError: anonymizeError(new Error("Bar")),
      })
    );

    expect(stringifyApolloError(apolloError)).toMatchSnapshot();
  });
});

describe("runServerSideQuery", () => {
  it("should return apollo result on success", async () => {
    const QUERY = gql`
      query {
        foo
      }
    `;
    const client = createApolloClient([
      {
        request: {
          query: QUERY,
        },
        result: {
          data: {
            foo: null,
          },
        },
      },
    ]);

    await expect(runServerSideQuery(client, QUERY)).resolves.toMatchObject({
      data: { foo: null },
    });
  });

  it("should re-raise GraphQL error", async () => {
    const QUERY = gql`
      query {
        foo
      }
    `;
    const client = createApolloClient([
      {
        request: {
          query: QUERY,
        },
        result: {
          data: {
            foo: null,
          },
          errors: [
            {
              locations: [
                {
                  column: 3,
                  line: 2,
                },
              ],
              message: "A foo error!",
              path: ["foo"],
            } as any,
          ],
        },
      },
    ]);

    await expect(runServerSideQuery(client, QUERY)).rejects.toThrowError(
      "GraphQL error: A foo error!"
    );
  });
});
