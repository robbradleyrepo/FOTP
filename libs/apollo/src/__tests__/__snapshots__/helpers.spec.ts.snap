// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sanitize should remove all \`__typename\` keys 1`] = `
Object {
  "bar": Object {
    "nested": Object {
      "value": 123,
    },
  },
  "baz": "baz",
  "foo": Array [
    Object {
      "index": 0,
    },
    Object {
      "index": 1,
    },
    Object {
      "index": 3,
    },
  ],
}
`;

exports[`stringifyApolloError should stringify GraphQL errors 1`] = `"{\\"graphQLErrors\\":\\"[\\\\\\"{\\\\\\\\\\\\\\"stack\\\\\\\\\\\\\\":\\\\\\\\\\\\\\"\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\"Dummy \`GraphQLError\` stack\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\"\\\\\\\\\\\\\\",\\\\\\\\\\\\\\"message\\\\\\\\\\\\\\":\\\\\\\\\\\\\\"\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\"Foo\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\"\\\\\\\\\\\\\\",\\\\\\\\\\\\\\"name\\\\\\\\\\\\\\":\\\\\\\\\\\\\\"\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\"GraphQLError\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\"\\\\\\\\\\\\\\"}\\\\\\"]\\",\\"networkError\\":\\"null\\",\\"message\\":\\"\\\\\\"GraphQL error: Foo\\\\\\"\\"}"`;

exports[`stringifyApolloError should stringify all top-level error data 1`] = `"{\\"graphQLErrors\\":\\"[]\\",\\"networkError\\":\\"null\\",\\"message\\":\\"\\\\\\"\\\\\\"\\"}"`;

exports[`stringifyApolloError should stringify network errors 1`] = `"{\\"graphQLErrors\\":\\"[]\\",\\"networkError\\":\\"\\\\\\"{\\\\\\\\\\\\\\"stack\\\\\\\\\\\\\\":\\\\\\\\\\\\\\"\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\"Dummy \`Error\` stack\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\"\\\\\\\\\\\\\\",\\\\\\\\\\\\\\"message\\\\\\\\\\\\\\":\\\\\\\\\\\\\\"\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\"Bar\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\"\\\\\\\\\\\\\\"}\\\\\\"\\",\\"message\\":\\"\\\\\\"Network error: Bar\\\\\\"\\"}"`;
